### PR TITLE
Optimize FusedBatchNormGrad.

### PR DESCRIPTION
**Pull request by codrut3**
Reuse the output buffer and allocate only one temporary tensor, when data format is NHWC and
GPU is used. This is based on the observation that cudnn can perform the backward computation
in place. The same idea is used in PR #15601 . 

This lowers GPU memory consumption and may improve performance, because fewer distinct memory addresses are accessed. It also permits a higher batch size.

I've also added a new test for the gradient computation.

### CODE REVIEW

**Comment by zhangyaobit** [comment]
@codrut3, thank you very much for the contribution! This is indeed a nice change. Unfortunately, as you noted, the change relies on un-documented/un-officially supported feature of cuDNN; the future cuDNN may change this behavior and would break users silently.

It seem un-safe to me to check in this change; we can certainly revisit this issue after cuDNN has official support on this.

Adding @zheng-xq

**Comment by zheng-xq** [comment]
Thanks for the contributions!

Adding @benbarsdell and @nluehr from NVIDIA here. How reasonable is it that future cudnn can follow this convention that backward batch-norm can happen in place?

Also for real models, such as resnet50 or inceptionv3, how much is the performance and maximum batch size improvement? If they are quite significant, we could condition this on cudnn-versions. But that is a lot of maintenance overhead, and we could only consider it if the improvement is sizable in real models. Thanks!

**Comment by codrut3** [comment]
Hi @zhangyaobit ,

I understand. If you think it's too much trouble I don't mind if you close the PR.

**Comment by nluehr** [comment]
In-place backward batch norm certainly works by accident rather than design. Also, execution can follow a variety of code paths depending on input size, GPU architecture, etc. I wouldn't be surprised if in-place breaks down in some corner cases. As noted above, it certainly could in future releases.

My assumption is that the memory savings should be modest. The out-of-place buffer is only needed during the actual cudnn call, and would be the size of one batch of activations. I would be interested to learn about any significant gains in important use cases. We might be able to add an explicitly in-place-safe variant.

**Comment by zhangyaobit** [comment]
Closing the PR for now. Please re-open if it turns out the issue needs to be re-visited.
